server/src/workspace/workspace-migration-runner/services/workspace-migration-enum.service.ts

```diff
@@ -0,0 +1,183 @@
+import { Injectable } from '@nestjs/common';
+
+import { QueryRunner } from 'typeorm';
+
+import { WorkspaceMigrationColumnAlter } from 'src/metadata/workspace-migration/workspace-migration.entity';
+
+@Injectable()
+export class WorkspaceMigrationEnumService {
+  async alterEnum(
+    queryRunner: QueryRunner,
+    schemaName: string,
+    tableName: string,
+    migrationColumn: WorkspaceMigrationColumnAlter,
+  ) {
+    const oldEnumTypeName = `${tableName}_${migrationColumn.columnName}_enum`;
+    const newEnumTypeName = `${tableName}_${migrationColumn.columnName}_enum_new`;
+    const enumValues =
+      migrationColumn.enum?.map((enumValue) => {
+        if (typeof enumValue === 'string') {
+          return enumValue;
+        }
+
+        return enumValue.to;
+      }) ?? [];
+
+    if (!migrationColumn.isNullable && !migrationColumn.defaultValue) {
+      migrationColumn.defaultValue = migrationColumn.enum?.[0];
+    }
+
+    // Create new enum type with new values
+    await this.createNewEnumType(
+      newEnumTypeName,
+      queryRunner,
+      schemaName,
+      enumValues,
+    );
+
+    // Temporarily change column type to text
+    await queryRunner.query(`
+      ALTER TABLE "${schemaName}"."${tableName}"
+      ALTER COLUMN "${migrationColumn.columnName}" TYPE TEXT
+    `);
+
+    // Migrate existing values to new values
+    await this.migrateEnumValues(
+      queryRunner,
+      schemaName,
+      tableName,
+      migrationColumn,
+    );
+
+    // Update existing rows to handle missing values
+    await this.handleMissingEnumValues(
```

Comment:
this is the first migration that will touch the data. I am very curious to see how it behaves on production.
On a regular project I would avoid that, but as we are doing migration per workspace, it's actually great. It will help migrating between versions of twenty handling data migration too and we will be able to test it on test workspaces, tier 3 workspaces before trying it on more important workspaces or clients!